### ISSUE TITLE
Make "joinRoom" language agnostic

### ISSUE THREAD

**Issue by adinapoli** [andyet/SimpleWebRTC#501]
Hey guys,

sorry if what follows is a silly question, feel free to close this ticket in that case 😉 

I have been using SimpleWebRTC to implement a prototype for a WebRTC application. The setup is a classic: a node.js server for the signalling part (partially ported from [this fork](https://github.com/azproduction/SimpleWebRTC-server)) and the [latest version](https://github.com/andyet/SimpleWebRTC/blob/master/latest-v2.js) of SimpleWebRTC on the client side. I always had in the back of my mind the curiosity to be able to swap the signalling server at whim, with something theoretically not even implemented in JavaScript. It seems doable for most parts, but looking at the [implementation of joinRoom](https://github.com/andyet/SimpleWebRTC/blob/master/simplewebrtc.bundle.js#L17715#L17750) seems like what it's doing is to get an optional callback from the client and emitting on the websocket a _serialised_ javascript function (the anonymous function at line 17718) which needs to be invoked on the signalling server side to actually trigger the initial handshake between peers.

This was empirically tested by me when I was puzzled about why this code (on the signalling server) was not triggering the handshake on the client side:

``` javascript
        client.on('join', function (roomName, callback) {
             // doing some ops
             // ...
            // NOTE: Not doing anything with callback
            // ...
        });
```

As soon as I actually triggered the callback calling `callback(null, roomDescription)` this worked,
as the `callback` here is not really the one the user passes from the client, but a lambda which will do the initial handshake _and then_, optionally, calling the user callback, which I think is a
bit confusing.

After this long preamble, here are my questions:

* Is the implementation of `joinRoom` really serialising and sending that anonymous function over the wire? If this is true, this basically shuts down any possibility for this library to be used
with a server written in something different than node, as the whole thing works only because 
the binary representation of such function is compatible with the server side. Is that really true or am I being totally mental here?

* Do you think it should be possible to avoid that callback at all, or at least modify the `joinRoom` implementation in a way that instead we serialise/deseriale some JSON or something language/platform agnostic instead of a JS function? I appreciate this will break virtually all the users of the library, and it should really be something for version `3.0.0`.

Sorry if what I'm writing doesn't make any sense, I'm mostly genuinely curious to see if I can
use SimpleWebRTC with a non-JS signalling server 😉  If the answer is yes, where should I look at? Documentation on the topic seems quite scarce.

Thank you in advance!

Alfredo

**Comment by fippo** [comment]
the callback isn't sent over the wire. You can easily check by looking at the network console in chrome and see the websocket frames.

We're using socket.io. Documentation on that is 'sparse' indeed.

**Comment by adinapoli** [comment]
Hey @fippo , thanks for the quick reply!

> the callback isn't sent over the wire. You can easily check by looking at the network console in chrome and see the websocket frames.

This is what I'm struggling to understand, maybe I'm ☕  depleted 😉 . It's clear to me that you guys are using `socket.io`, but this is my understanding of the current implementation (which I'm sure I've got wrong!):

* `prototype.joinRoom` has this line in it:

``` javascript
this.connection.emit('join', name, function (err, roomDescription) {
```

According to the implementation of [emit](https://github.com/socketio/socket.io-client/blob/master/lib/socket.js#L134) on the socket.io client-side implementation (which is what simplewebrtc should be using, I guess), they seems to be emitting to the socket a new [binary event](https://github.com/socketio/socket.io-client/blob/master/lib/socket.js#L142) as the check `hasBin(args)` returns true, as args would contain, at this point in time `["join", name, <function>]`. At this point the control passes on the _server_ which would listen on the socket and [react to the event](https://github.com/socketio/socket.io/blob/master/lib/socket.js#L303) according to which packet type we got from the client. This to me seems indeed that the binary blob travels in the socket, and as the two ends of the sockets can be geographically distant, this is what I intended for "over the wire". The last function I linked does a switch on the packet type (in this case would be `BINARY_EVENT` and would call [onevent](https://github.com/socketio/socket.io/blob/master/lib/socket.js#L338), which seems to do two things:

* Reify the binary blob [back into a JS data structure](https://github.com/socketio/socket.io/blob/master/lib/socket.js#L339)

* Applying the function [calling emit.apply](https://github.com/socketio/socket.io/blob/master/lib/socket.js#L347) 

Unless I'm mistaken this seems to me as the library (socket.io) is essentially passing everything, including JS functions from the client to the server, ultimately applying it. If this was true (I bet I'm being naive here!) this means the entire library is coupled with the JS environment and assumes the binary blob can be interpreted corrected by the server side.

But this cannot be true, as I have seen there as alternative client implementations (example a Java/Android version) so surely the Socket.IO devs should have thought about such coupling.

What do you guys think? Am I essentially performing mental masturbation here? 😬 

Thank you!

Alfredo

**Comment by fippo** [comment]
no, I think what happens is in https://github.com/socketio/socket.io-client/blob/master/lib/socket.js#L148-L153

**Comment by adinapoli** [comment]
Hey @fippo , thanks for replying!

That piece of code you linked simply modify the structure of the packet, namely modify the `acks` array and setting the `packet.id` to something else, but I think that eventually a packet of this shape `var packet = { type: parserType, data: args };` is pushed down the wire with this piece of code:

```
  if (this.connected) {
    this.packet(packet);
  } else {
    this.sendBuffer.push(packet);
  }
```

What am I missing?

**Comment by fippo** [comment]
right but the function is popped from args

**Comment by adinapoli** [comment]
> right but the function is popped from args

aha! That's the bit I was missing, thank you! Sometimes when working with immutable data structures all day long it takes a while for your brain to realise the `args` array gets modified along the way 😁 

Now I'm with you when you say that the function is not sent over the wires! Probably the last thing I'm struggling to understand is how come, then, the very same callback function we declare on the client side "magically appears" on the server side (in the call `on("join", function(roomName, callback)`) to the point that if I omit calling `callback` explicitly the whole handshake process grind to an halt. I suspect socket.io's acking mechanism is somewhat involved, but my socket.io-fu is somewhat limited. Do you guys have any clue?

Just to make things clear, mine is a genuine question and I'm not trying to subtly insinuate something different here 😉  Thanks for you time guys, it's very appreciated!

**Comment by fippo** [comment]
no idea but thanks for making me look, I learned a bit about socket.io ;-)

**Comment by adinapoli** [comment]
> no idea but thanks for making me look, I learned a bit about socket.io ;-)

Hehe you're welcome, but I suspect  this is mutual, I had fun in poking around with the socket.io implementation 😉  And thanks for the work on this library guys, so far has been a real pleasure to use, and I hope to be able to write my own signalling server in Haskell if I will ever be able to get at the bottom of the callback story 😸
